### PR TITLE
Optimise MapBlockMesh functions

### PR DESCRIPTION
**Pull request by gregorycu**
*** Please put feedback here, not on IRC. Thank you. ***
MapBlockMesh::MapBlockMesh
MapBlockMesh::getTileInfo
MapBlockMesh::makeFastFace
MapBlockMesh::getSmoothLightCombined
Speedup is about 33% for these functions.

### CODE REVIEW

File: src/voxel.h [hunk 1]
```diff
@@ -413,10 +413,21 @@ class VoxelManipulator /*: public NodeContainer*/
...
 	{
 		return m_data[m_area.index(p)];
 	}
+
+	MapNode & getNodeRefUnsafeCheckFlags(const v3s16 &p)
+	{
+		s32 index = m_area.index(p);
+
+		if (m_flags[index] & VOXELFLAG_NO_DATA)
+			return ContextIgnoreNode;
```
Comment (Zeno-) [inline src/voxel.h]: Just commenting on what I said on IRC.

This is not safe (and not desirable, IMO). The issue is that the static member `ContextIgnoreNode` is returned as a non-const object reference. Therefore anything or anyone using this function can change `ContextIgnoreNode` and cause mayhem.

Comment (gregorycu) [inline src/voxel.h]: I'll take a look.

Comment (gregorycu) [inline src/voxel.h]: I have amended the commit to make the function const

**Comment by Zeno-** [comment]
In the current state it's in? :-1:

**Comment by Zeno-** [comment]
Marking as WIP because the author refuses to discuss the PR or answer questions in IRC.

E.g.
```
<Zeno`> why is MapNode VoxelManipulator::ContentIgnoreNode = MapNode(CONTENT_IGNORE);  extern (and not static) and why isn't it const?

<hmmm> can't you make ContentIgnoreNode static const if you initialize it with brackets?
```

See: http://irc.minetest.ru/minetest-dev/2015-01-26#i_4131662 (and the rest of the day's logs)

**Comment by gregorycu** [comment]
With regards to the claim that I refuse to discuss the PR or answer questions in IRC, looking at the logs, I clearly responded to hmmm and discussed this issue.

**Comment by gregorycu** [comment]
I have made the class static member const as requested

**Comment by Zeno-** [comment]
Closing. Author will not discuss the PR on IRC and therefore questions about it cannot be asked.

**Comment by gregorycu** [comment]
What was the technical reason why this was closed?

**Comment by Zeno-** [comment]
Yes. And it's already static. But it cannot be discussed.

**Comment by gregorycu** [comment]
What are the outstanding issues with this PR?

**Comment by Zeno-** [comment]
The outstanding issues are that you, for no apparent reason, will not discuss it on IRC.

**Comment by gregorycu** [comment]
if there are outstanding technical issues, please let them be known here, where everyone can see them. Otherwise I request that you reopen this PR so that others may give their feedback, and possibly pull it.

**Comment by Zeno-** [comment]
There are no technical issues. There are social issues.

**Comment by PilzAdam** [comment]
WTF @Zeno-

**Comment by Zeno-** [comment]
@PilzAdam see IRC logs

**Comment by gregorycu** [comment]
Please don't policitise my PR. As far as I can gather, there are no questions regarding this PR. As per @celeron55 's request, I have marked the PR as requesting feedback on Github.

**Comment by Zeno-** [comment]
You choosing not to converse with me on IRC (for no reason that I can comprehend) is just silly -- the question I asked could have been answered in one very short sentence. Anyway, I will reopen this PR but do not support it.

Do you want me to paste yesterday's IRC logs here?

**Comment by gregorycu** [comment]
No thank you, unless the logs in their entirety highlight some issue I am not aware of. I am aware of hmm and you asking the member to be made const. Which I have done (and I stated that I did).

On the record of silly, the fact that you closed a PR because I wasn't talking to you on IRC (even though there was nothing to talk about) suggests to me that I am vindicated in ignoring you on IRC.

**Comment by Zeno-** [comment]
@gregorycu 
a) You've attacked me on IRC for reasons that escape me (personal attacks without provocation)
b) You've ignored me on IRC for no apparent reason and when asked why by other coredevs you've refused to give a sensible answer other than what boils down to "I want to"
c) You were explicitly asked **not** to ignore me in #minetest-dev by more than one coredev (including celeron55)

I don't have the slightest clue why I am ignored by you on IRC and you won't explain why. I have no idea why caused your sudden aggressive behavior and personal attacks on IRC. 

The reason this was closed:
On IRC, I asked you several questions watched your conservation for a while and repeated the main question several time. I then said (while you were active) I would close the issue if you cannot or would not answer the question. I gave you 30 minutes or so to respond and then reiterated that the PR would be closed if there was no response provided and I asked the question again. There was no response so I (with due warning and process) closed the PR.

**Comment by gregorycu** [comment]
Just to give you the closure I think you're looking for:
a) I never attacked you, although I did jokingly ask if your bias against C++ was because you were old. I apologise for that.
b) We would talk in friendly terms, but at the drop of a hat you would get defensive and act offended. I don't want to spend my time on IRC defending innocent talk. It's easier to ignore you, and instead do all PR related conversation in GitHub. I didn't think this would be a problem.
c) I was asked by celecron55, and upon further discussion he said it would be fine provided I was explicit in my PRs that feedback is invited in GitHub

You are a vindictive person willing to use a project you were entrusted with to push a personal grievance, I've tried to keep this PR professional, but you insisted on closing it, even after it was reopened by other dev and even after celeron55 said it wouldn't be a problem. You continually stated that I wasn't answering questions, but you then went on to say that there were no outstanding questions.

There are numerous mischaracteristations made by you, about me. For example, the assertion that I was refusing to answer questions and then posting a question from hmmm: "<hmmm> can't you make ContentIgnoreNode static const if you initialize it with brackets?". A question which I immediately answered. In fact, there was a discussion about it. "When gregorycu can converse with others when he is on IRC instead of ignoring their questions it can be re-opened." What others? I was just ignoring you.

Basically you made something that was an non-issue, and personal, and tried to make out that I was ignoring everybody and against the whole channel.

You also went on to attack my work, for no good reason, disputing figures even though you had no reason to suspect they were wrong. You even attempted to downplay an optimisation from 28 seconds down to half second, trying to manipulate figures to make is seem like an increase.

Addendum Response:

I didn't even feel a twinge of anger about the deadlock, the discussion about C++11 wasn't personal, you the arguments you were making were not persuasive at all. Basically the conversation went on, and, I think, when you realised you didn't have a leg to stand on, you screamed victim and personal attacks. It was then i decided you were not worth conversing with on a casual basis.

My email is gregory.currie at gmail.com. If you have anything else to say, you can say it there. I think I've answered all your implied questions to death now, and looking forward to this PR getting back on-track.

**Comment by Zeno-** [comment]
The only reason it was closed was because I asked if it should be closed and said I would close it if there was no comment. There was no response. This is not being vindictive.

**Comment by gregorycu** [comment]
And when I asked it to be reopened? I guess you just didn't see it.

**Comment by Zeno-** [comment]
> And when I asked it to be reopened? I guess you just didn't see it.

I responded to that when you asked it in IRC (within 5 minutes of you asking).

Anyway, I don't want to keep this discussion going.

**Comment by gregorycu** [comment]
I'm going to let you have the last say, I'm going to withdraw this PR, and reissue another without the crummy comments.

**Comment by Zeno-** [comment]
That's ok. Probably a good idea as well because if a commit cannot be discussed on IRC it probably won't be merged for quite a while (for a commit like this 2 core devs need to agree). The code is good (based on a quick glance) now that the issues have been addressed.

A fair majority of your previous PRs were merged by me, and of those I corrected things like whitespace errors and merge conflicts. I even helped you fix your messed up local repo and helped you rebase it. To say I didn't encourage you and give every possible assistance in getting your PRs into master is... I dunno... innaccurate.

**Comment by Zeno-** [comment]
For reference: http://irc.minetest.ru/minetest-dev/2015-01-27#i_4133409  (and surrounding lines, of course)

**Pull request closed without merging**